File: Functional programming - OCaml/laboratory tasks/List4.ml
(*Zadanie 0*)
let log = fun prefix -> fun datetime -> fun text ->
  print_string ("[{"^prefix^"}] {"^datetime^"} \t {"^text^"} \n")
;;

(*Zadanie 1*)
let map list f = 
  let rec mapIn list newList f =
    match list with
    | [] -> newList
    | h::t -> mapIn t (newList@[f h]) f 
  in mapIn list [] f
;;

(*Zadanie 2 *)
let  filter list f = 
  let rec filterIn list newList f =
    match list with
      | [] -> newList
      | h::t -> 
       match f h with
       | true -> filterIn t (newList@[h]) f
       | false -> filterIn t newList f
  in filterIn list [] f
;;

(*Zadanie 3*)
let rec reduce list acc f =
  match list with
  | [] -> acc
  | h::t -> reduce t (f acc h) f
;;

(*Zadanie 4*)
let mean list = 
  (List.fold_left (fun x y -> x +. y) 0. list) /. (List.fold_left (fun x y -> x +. 1.0) 0. list)
;;

(*Zadanie 5*)
let toList str =
  let rec toListIn str list i length =
    if i >= length then list
    else toListIn str (list@[str.[i]]) (i+1) length
  in toListIn str [] 0 (String.length str)  
;;

  let toString list =
    let s = "" in
    let rec toStringIn str list =
      match list with
      | [] -> str
      | h::t -> toStringIn (String.concat str [""; String.make 1 h]) t
    in toStringIn s list
;;

let akronym str =
  let rec newReduce list acc f wasSpace=
    match list with
    | [] -> acc
    | h::t -> 
      match h with
      | ' ' -> newReduce t (acc) (f) (true)
      | _ -> 
        if wasSpace then newReduce t (f acc h) (f) (false)
        else newReduce t (acc) (f) wasSpace
  in
  let addFirstLetter acc h = String.concat acc [""; String.make 1 h] in
  newReduce (toList str) ("") (addFirstLetter) (true)
;;

(*This method is an alternative way of solving Task 5 
but is based on mutable variables and uses more variables than the above solution.*)
let alternativeAkronym str =
  let space = ref 0  (*Variable told as if last char was space (then its 0) else its 1*) in
  let wasSpace = ref 0 (*Variable will be only used to compare with variable space*) in
  let isThisFirstLetterOfWord c =
    let isThisFirstLetterOfWordIn c =
      match c with
    | ' ' -> space := 0; false
    | _ -> if space = wasSpace then space := 1; true
    in
    match c with
    | ' ' -> isThisFirstLetterOfWordIn c
    | _ -> 
      if space = wasSpace then isThisFirstLetterOfWordIn c
      else false
  in 
  toString (filter (toList str) isThisFirstLetterOfWord)
;;



(*Zadanie 6*)
let sqrListSmallerThenSum list =
  let sum = List.fold_left (fun x y -> x+.y) 0.0 list in 
  let isCubeSmallerThenSum x = if x*.x*.x <= sum then true else false in
  (List.filter (isCubeSmallerThenSum) list )
;;  

File: Functional programming - Scala/Laboratory tasks/List4/src/main/scala/List4.scala
import scala.collection.View.Filter

object List4 {

   //Zadanie 1
   def map[A](list : List[A], f : A => A): List[A] = {
      def mapIn[A](list: List[A], newList: List[A], f: A=>A): List[A] ={
         list match {
            case Nil => newList
            case h :: t => mapIn[A](t, newList:::List(f(h)), f)
         }
      }
      mapIn[A](list, Nil, f)
   }

   //Zadanie 2
   def filter[A](list: List[A], f: A => Boolean): List[A] = {
      def filterIn[A](list : List[A], newList : List[A], f: A => Boolean): List[A] = {
         list match {
            case Nil => newList
            case h :: t =>
               f(h) match {
                  case false => filterIn[A](t, newList, f)
                  case true => filterIn[A](t, newList:::List(h), f)
               }
         }
      }
      filterIn[A](list, Nil, f)
   }

   //Zadanie 3
   def reduce[A](list : List[A], acc : A, f: (A, A) => A): A = {
      list match {
         case Nil => acc
         case h::t => reduce[A](t, f(acc, h), f)
      }
   }

   def sqr(x : Int): Int = {x*x}
   def smallerThen3(x : Int): Boolean = {
      if x < 3 then true
      else false
   }
   def sum(x : Int, y : Int): Int = {x+y}

   //Zadanie 4
   def mean(list: List[Int]): Float = {
      ( (list foldLeft 0)((x, y) => x + y) ) / ((list foldLeft 0)((x, y) => x + 1))
   }

   //Zadanie 5
   def akronym(str : String): String = {
      var wasSpace = true
      def isThisFirstLetterOfWord(c : Char): Boolean = {
        c match {
           case ' ' =>
              wasSpace = true
              return false
           case _ =>
              if(wasSpace){
                 wasSpace = false
                 return true
              }
              else false
        }
      }
      str.filter(isThisFirstLetterOfWord)
   }

   //Zadanie 6
   def sqrSmallerThenX3(list: List[Int]): List[Int] = {
      val sum = (list foldLeft 0)((x, y) => x + y)
      def isCubeSmallerThenSum(x : Int): Boolean = {
         if x*x*x <= sum then true
         else false
      }
      filter[Int](list, isCubeSmallerThenSum)
   }

   @main def main(): Unit = {
      println(map[Int](List(1, 2, 3, 4, 5), sqr))
      println(filter[Int](List(1, 2, 3, 4, 5), smallerThen3))
      println(reduce[Int](List(1, 2, 3, 4, 5), 0, sum))
      println(mean(List(1, 2, 3, 4, 5)))
      println(akronym("Zakład Ubezpieczeń Społecznych"))
      println(sqrSmallerThenX3(List(1, 2, 3, 4, 5)))
   }
}
